### ISSUE TITLE
Unable to correctly implement State Caching Proxy Pattern using dbusutil

### ISSUE THREAD

**Issue by patrickmeiring** [godbus/dbus#187]
Hi All,

*Background:*
I am working on an application that involves tracking the state of a DBus object. The object I am using provides a method to get its current state, as well as a signal to notify us whenever that state changes.

I would like to build a local object which tracks the state of the remote object and caches it for quick access. In the literature, I have seen this referred to as a a _State Caching Proxy pattern_. 

The guarantee I would like to achieve is that once the object has been initialised, the observable states of this local object will be a contiguous subsequence of the observable states of the remote object. That is, if the remote object transitions through states S(0), S(1) ... S(n) during the lifetime of the system, the local object will expose some contiguous subsequence of those states (for as the local object exists and is subscribed). 
Valid observation trace: S(3), S(4), S(5), S(6) ...
Invalid observation trace: S(4), S(3), S(5), S(6) ... (States received out of order.)
Invalid observation trace: S(3), S(5), S(6). (Missing state changes.)

If at the point the local caching proxy object is initialised, it exposes state S(3), I expect to be able to observe all future state changes (i.e. S(4), S(5), S(6), ..) in exactly that order without missing any changes *and* without any other state changes mixed in. 

The State Caching Proxy pattern is described in section 5.2 of Olli Salli's masters thesis:
"Building object-oriented software with the D-Bus messaging system"
https://www.utupub.fi/bitstream/handle/10024/78669/gradu2012Salli.pdf?sequence=1

As described in that thesis, the correct way of implementing this pattern is to subscribe for changes first and _then_ ask for the object's current state. Only once the current state is received, is the object considered initialised and are further state changes used.

*Problem*
Unfortunately, dbusutil presents two impediments to correctly implementing this pattern:
- Signals are not necessarily received in-order. The current implementation spins up a separate goroutine to dispatch each signal. There is no guarantee these goroutines will execute in-order (indeed, in our observation, for signals sent in close succession, the goroutines execute in arbitrary order. As such, we are receiving signals out of order.). 
- I have no way of identifying if a signal was received before or after the method reply that told us the initial state of the object. Both signals and method responses are handled in a somewhat asynchronous manner, using different channels and goroutines. Consequently, the original order of messages on the DBus is lost. This ordering is the only way in which we can ascertain if a state change signal is for an old state (prior to us querying the initial state) and should be discarded, or is for a new state.

To my knowledge, other D-Bus libraries such as GDBus, provide mechanisms to correctly implement this pattern by offering synchronous processing of D-Bus messages.

I would be open to contributing changes to fixing dbusutil to allow the above pattern, but it seems to me that it may require significant architectural changes (mostly for the second issue: exposing ordering between method responses and signals). I would therefore like to enquire what the maintainers of this package think about this issue.

Thanks in advance,
Patrick

**Comment by jsouthworth** [comment]
We have the 'Handler' and 'SignalHandler' interfaces which should allow you to override the default behavior to achieve what you want. They may be used as part of the ConnOption lists to the Dial/Connect functions. If there is something that needs changing that cannot be done using these methods a pull request would be welcomed for discussion. We have these to avoid breaking clients using the current semantics (broken as they may be). I use these in my own client projects to avoid signal race issues.

**Comment by jsouthworth** [comment]
I'd also be open to including additional implementations of those handlers as part of this library so that a client could choose the desired behavior. The default would still need to be left as it is now.

**Comment by patrickmeiring** [comment]
Thanks for the reply @jsouthworth.

I can see how using a different signal handler would allow one to make sure signals are always delivered in the order they are received on DBus. (I.E. Solve problem 1.). 

However, I do not think this is sufficient to solve problem 2, that is: giving the application logic the ordering between method replies and signals on the dbus connection. 

I think there are two ways this ordering information can be provided:
1. Implicitly, via program control flow. That is, we block processing of the next incoming signal or method reply until the previous signal or method reply has been processed by user code. User code can then rely on the guarantee that their processing will run in the order the messages were received on the DBus. This approach would rely on user code telling us when it has processed an event, or providing us a synchronous method to do the processing; where returning from the method indicates processing is complete.
2. Explicitly, via sequence number that is consistent between signals and method replies. Here, we simply number the signals and method replies as they come in and let user code worry about processing them in order, if that matters.

Option 1 might be the most convenient for user code interfacing with the library, however, this is likely to require significant re-architecture of this library. For one, note that all processing of signals and method replies happens asynchronously to their being received:
1. It is currently common practice to buffer signals into channels, and have goroutines asynchronously process them. For example, NewSignalWatcher buffers signals in a channel called allSigs (written to by the signal handler) while a goroutine places the matching ones in another channel calls Signals. There is an unspecified delay between when a signal appears in allSigs and when it becomes visible to user code in Signals, attributable to goroutine scheduling. There is a further unspecified delay until user code reads the Signals channel and does its processing, attributable also to goroutine scheduling.
2. The processing of method replies is asynchronous to their receiving. Once a response is received, object.CallWithContext will return, but this occurs in a seperate goroutine to the inWorker that received the method reply. There is an unspecified delay between when .CallWithContext returns and when the information is meaningfully processed; and user code does not tell us when it is done processing the result.

This means switching to sequential processing would likely require large-scale changes in the library. Either:
- One would switch to synchronous processing; replacing uses of channels with delegates, where the delegate returning indicates processing is complete.
- One would keep asynchronous processing, but block further processing of messages until user code has notified us (asynchronously) that processing is complete (e.g. by waiting until user code calls some method .continueProcessing()). It would require significant discipline on the user side, and we would need to be careful that .continueProcessing() is called once and exactly once for each received method reply and signal.

Option 2. This option is fairly self-explanatory. Given signals/method replies are numbered according to a common counter, one can in post-processing choose to ignore state change signals with a sequence number below that of the method reply indicating the initial state.

**Comment by patrickmeiring** [comment]
Based on the above, my proposed changes would be to:
1. Number signals and method replies as they come in with a sequence number from a common counter.
2. Add a signal handler that ensures signals are sent to user-code in the order they are received. 

The psudocode for the race-free solution to this problem would then be:
```golang
watcher := NewSignalWatcher("StateChange")
initialState, seq := GetInitialState()
for {
    select {
          case signal<-watcher.Signal:
              if (signal.Seq > seq) { 
                    output("State changed to: " + event.State)
              }
    }
}
```

While (1) alone might be sufficient to solve the problem articulated in this thread, I think (2) has significant value in isolation, and also avoids the issue of progress associated with (1). That is, if one received:
Sequence Number | Signal
1 | Name changed to "A"
3 | Name changed to "B"
one would not know how long to wait for a potential event with sequence number 2 still making its way through the plumbing. If events are guaranteed to arrive in order, this is perfectly clear, and the signal handler can always make progress (like in the psudocode above).

**Comment by patrickmeiring** [comment]
If you don't mind, I'm going to use this pull request as a bit of a development branch to quickly try out a few changes and debug what is going wrong on PowerPC architectures. When I'm happy with the changes, I will create a new pull request and discard this one.
